### PR TITLE
add eslint rules

### PR DESCRIPTION
**Pull request by rezaerami**
- add new rules to help with code style
- add prettier and prettier rc to sync IDE with project rules
- fix code style of files with new standard

### CODE REVIEW

File: packages/apps/transfer/.eslintrc.js [hunk 1]
```diff
@@ -1,8 +1,32 @@
+const fs = require('fs');
+
+const prettierOptions = JSON.parse(fs.readFileSync('./.prettierrc', 'utf8'));
```
Comment (rezaerami) [inline packages/apps/transfer/.eslintrc.js:3]: importing prettier this way helps to sync IDEs such as webstorm which has it's own formatted with project standards
cause now both IDE and linter will read prettier config from same config file

File: packages/apps/transfer/.eslintrc.js [hunk 2]
```diff
@@ -1,8 +1,32 @@
...
 
 module.exports = {
-  plugins: ['jest'],
+  plugins: ['prettier', 'jest'],
   extends: ['next/core-web-vitals', '@kadena-dev/eslint-config/profile/react'],
   parserOptions: { tsconfigRootDir: __dirname },
   ignorePatterns: ['**/*.spec.ts', '**/*.spec.tsx', '**/generated/**'],
+  rules: {
+    'prettier/prettier': ['error', prettierOptions],
+    '@typescript-eslint/strict-boolean-expressions': 0,
```
Comment (rezaerami) [inline packages/apps/transfer/.eslintrc.js:13]: disable boolean expressions

File: packages/apps/transfer/.eslintrc.js [hunk 3]
```diff
@@ -1,8 +1,32 @@
...
+    'arrow-body-style': [2, 'as-needed'],
```
Comment (rezaerami) [inline packages/apps/transfer/.eslintrc.js:14]: simplify arrow function and use parentheses instead of return keyword when function only returns 
e.g.
```
// turns this
const foo = () => {
    return "bar";
}
// to this
const foo = () => "bar"
```

File: packages/apps/transfer/.eslintrc.js [hunk 4]
```diff
@@ -1,8 +1,32 @@
...
+    'comma-dangle': [2, 'always-multiline'],
```
Comment (rezaerami) [inline packages/apps/transfer/.eslintrc.js:15]: adds extra comma to last item of an object or array if it's multiline
```
const foo = [
    "foo",
]
```

File: packages/apps/transfer/.eslintrc.js [hunk 5]
```diff
@@ -1,8 +1,32 @@
...
+    'prefer-template': 2,
```
Comment (rezaerami) [inline packages/apps/transfer/.eslintrc.js:16]: prefers to use templates instead of string concatenates
```
// turns this
const foo = "hi" + name;
// to this
const foo = `hi ${name}`;
```

File: packages/apps/transfer/.eslintrc.js [hunk 6]
```diff
@@ -1,8 +1,32 @@
...
+    'import/newline-after-import': 1,
+    'import/no-unresolved': 2,
```
Comment (rezaerami) [inline packages/apps/transfer/.eslintrc.js:18]: this is super useful especially for windows users with a repo manager
since we have a rush, on windows we'll have root node modules, therefore if a package is not listed in the package.json of one of the packages, it'll allow you to use it since you have it in root, but then on deployment it figures it out and it.
but with this rule, if you import a package, it has to be listed in package.json

this issue happens often on windows machines

File: packages/apps/transfer/.eslintrc.js [hunk 7]
```diff
@@ -1,8 +1,32 @@
...
+    'jsx-a11y/aria-props': 2,
```
Comment (rezaerami) [inline packages/apps/transfer/.eslintrc.js:19]: JSX-ally rules are for accessibility 
forcing user to use correct roles in case they used aria-*, also they should be valid regarding the element, for example an anchor IF had aria props, then it should have role of a link

File: packages/apps/transfer/.eslintrc.js [hunk 8]
```diff
@@ -1,8 +1,32 @@
...
+    'jsx-a11y/role-has-required-aria-props': 2,
+    'jsx-a11y/heading-has-content': 1,
+    'jsx-a11y/mouse-events-have-key-events': 2,
```
Comment (rezaerami) [inline packages/apps/transfer/.eslintrc.js:22]: to improve accessibility for for users without a mouse or touchpad this is also really useful, it forces you to have the same value for onFocus and onBlur, in case you used onMouseOver and onMouseOut

File: packages/apps/transfer/.eslintrc.js [hunk 9]
```diff
@@ -1,8 +1,32 @@
...
+    'jsx-a11y/role-supports-aria-props': 2,
+    'react/function-component-definition': [
```
Comment (rezaerami) [inline packages/apps/transfer/.eslintrc.js:24]: prefers to have arrow functions for components

File: packages/apps/transfer/.eslintrc.js (hunk 3 above)
Comment (sstraatemans) [reply packages/apps/transfer/.eslintrc.js:14]: hm would prefer to be free in your choice with this

**Review by alber70g — CHANGES_REQUESTED** [review]
Let's move all the rules to a next-specific config in the @kadena-dev/eslint-config project.

Also, as per conversation, we can remove some rules _only if_ it's possible to have those rules removed for `jsx` parts only. Anything not-ui related should adhere to the stricter ruleset as defined in the profiles currently

**Comment by webpro** [comment]
> Let's move all the rules to a next-specific config in the @kadena-dev/eslint-config project.

Perhaps "react" or "jsx" or "ui" or "loose" is more targeted/clear. Applies to UI logic, design system and React/JSX components in general.

> Also, as per conversation, we can remove some rules _only if_ it's possible to have those rules removed for `jsx` parts only. Anything not-ui related should adhere to the stricter ruleset as defined in the profiles currently

Maybe it could be as simple as applying the new rules to all `.tsx` files?

**Comment by rezaerami** [comment]
> 

yep it's totally possible tolimit certain rules to certain file formats
e.g.
```

  "overrides": [
    {
      "files": ["*.jsx"],
      "rules": {
        "@typescript-eslint/strict-boolean-expressions": "off"
      }
    }
  ],
  ```

**Comment by rezaerami** [comment]
> Maybe it could be as simple as applying the new rules to all .tsx files?
well, some of them such as arrow function prettier and import ones are not tsx-only
but indeed, we can apply some rules such as jsx ally ones only to jsx files, although when you have jsx code IDE will suggest you to convert it to .tsx | .jsx

therefore if there is no jsx code, those rules won't apply, and if there is, then the file extension will be tsx | jsx for sure

**Comment by alber70g** [comment]
We discussed this offline and @rezaerami will move most of the rules to the `@kadena-dev/eslint-config`, to either the `lib`, `react` or `next` config.

**Review by MRVDH — APPROVED** [review]

File: packages/apps/transfer/.eslintrc.js [hunk 10]
```diff
@@ -2,7 +2,7 @@ require('@rushstack/eslint-config/patch/modern-module-resolution');
 
 module.exports = {
   plugins: ['jest'],
-  extends: ['next/core-web-vitals', '@kadena-dev/eslint-config/profile/react'],
+  extends: ['@kadena-dev/eslint-config/profile/next'],
   parserOptions: { tsconfigRootDir: __dirname },
-  ignorePatterns: ['**/*.spec.ts', '**/*.spec.tsx', '**/generated/**'],
+  ignorePatterns: ['**/generated/**'],
```
Comment (alber70g) [inline packages/apps/transfer/.eslintrc.js:7]: Does removing this still work with VSCode ESLint Plugin? I had issues with the base config not being able to resolve the plugins from next

File: packages/apps/transfer/src/pages/transfer/cross-chain-transfer-tracker/index.tsx [hunk 11]
```diff
@@ -1,12 +1,10 @@
...
-    <MainLayout title="Kadena Coin Transfer">
-      <h3>TO DO: Cross Chain Transfer Tracker</h3>
-    </MainLayout>
-  );
-};
+const CrossChainTransferTracker: FC = () => (
+  <MainLayout title="Kadena Coin Transfer">
+    <h3>TO DO: Cross Chain Transfer Tracker</h3>
+  </MainLayout>
+);
```
Comment (alber70g) [inline packages/apps/transfer/src/pages/transfer/cross-chain-transfer-tracker/index.tsx:8]: It seems like we have decided to keep the code as it is after our discussion.

I.e.: when its written like `const x = () => { return "" }` it would **not** be rewritten to `const x = () => ""`

File: packages/apps/transfer/src/utils/env.ts [hunk 12]
```diff
@@ -11,6 +11,8 @@ export interface IEnvInterface {
   DEFAULT_SENDER?: string;
 }
 
+type IRequiredEnv = Required<IEnvInterface>;
```
Comment (alber70g) [inline packages/apps/transfer/src/utils/env.ts:14]: types shouldn't be prefixed with `I`

File: packages/libs/react-components/package.json [hunk 13]
```diff
@@ -57,6 +57,7 @@
     "@types/node": "^16.0.0",
     "@types/react": "^18.0.30",
     "eslint": "^8.15.0",
+    "eslint-import-resolver-alias": "~1.1.2",
```
Comment (alber70g) [inline packages/libs/react-components/package.json:60]: Why is this added? Can't we add it to the root config? Or will that not work with IDEs?

File: packages/tools/eslint-config/package.json [hunk 14]
```diff
@@ -27,6 +27,10 @@
   "devDependencies": {
     "@rushstack/eslint-config": "~3.3.0",
     "eslint": "^8.15.0",
+    "eslint-config-prettier": "~8.8.0",
+    "eslint-plugin-import": "^2.26.0",
+    "eslint-plugin-jsx-a11y": "~6.7.1",
+    "eslint-plugin-prettier": "~4.2.1",
```
Comment (alber70g) [inline packages/tools/eslint-config/package.json:33]: Will adding them to `devDependencies` only work when this config is used in other packages? Do those packages need to add those dependencies themselves, or can we move this to `dependencies` and will it work as well? Or will it _not_ anyway, whether we put them as regular- _or_ devdependencies?

File: packages/tools/eslint-config/profile/lib.js [hunk 15]
```diff
@@ -1,3 +1,10 @@
+const fs = require('fs');
+const path = require('path');
+
+const prettierOptions = JSON.parse(
+  fs.readFileSync(path.resolve(__dirname, '../../../../.prettierrc'), 'utf8'),
+);
+
```
Comment (alber70g) [inline packages/tools/eslint-config/profile/lib.js:7]: Can we add a comment that this is specific to our monorepo where, according to `rush.json#projectFolderMin/MaxDepth`, a package will **always** be in the 3rd nested directory (`packages/{apps,libs,tools}/package-name/*`). 

Should we also add to the comment that a more generic alternative could be to use:

```js
const prettierrcPath = `${require('child_process')
  .execSync('git rev-parse --show-toplevel')
  .toString()
  .replace(/\n/, '')}/.prettierrc`;
```

File: packages/tools/eslint-config/profile/lib.js
Comment (alber70g) [inline packages/tools/eslint-config/profile/lib.js:1]: Just noticed there's a ESLint rule for `sort-package-json`. Maybe we want to add it as an eslint rule instead and remove the package?

https://github.com/kellyselden/eslint-plugin-json-files#supported-rules:~:text=json%2Dfiles/sort%2Dpackage%2Djson

File: packages/tools/eslint-config/profile/react.js [hunk 16]
```diff
@@ -1,19 +1,41 @@
...
+      },
+    ],
   },
   settings: {
     react: {
       version: 'detect',
     },
   },
+  // turned off boolean expressions for jsx and tsx files, so we can have non-boolean statements in react components
+  // this rule should be enabled for js and ts files even in a ReactJS or NextJS setup
```
Comment (alber70g) [inline packages/tools/eslint-config/profile/react.js:30]:
```suggestion
  // disable strict-boolean-expressions for jsx and tsx files
  // this allows for truthy/falsy statements in react components
  // this rule must be enabled for js and ts files even in a ReactJS or NextJS setup
```

File: packages/tools/kda-cli/.eslintrc.cjs [hunk 17]
```diff
@@ -4,4 +4,7 @@ require('@rushstack/eslint-config/patch/modern-module-resolution');
 module.exports = {
   extends: ['@kadena-dev/eslint-config/profile/lib'],
   parserOptions: { tsconfigRootDir: __dirname },
+  rules: {
+    'import/no-unresolved': 'off',
+  },
```
Comment (alber70g) [inline packages/tools/kda-cli/.eslintrc.cjs:9]: Why do we have this? If there's a legitimate reason, let's document that in a comment

**Review by alber70g — CHANGES_REQUESTED** [review]
Some questions and additions

File: packages/apps/transfer/src/pages/transfer/cross-chain-transfer-tracker/index.tsx (hunk 11 above)
Comment (rezaerami) [reply packages/apps/transfer/src/pages/transfer/cross-chain-transfer-tracker/index.tsx:8]: ah no, this is coming from previous linter when I enabled that rule, now it's not enabled so it won't effect anything, this is just an old change

File: packages/apps/transfer/src/utils/env.ts (hunk 12 above)
Comment (rezaerami) [reply packages/apps/transfer/src/utils/env.ts:14]: fixed

File: packages/tools/eslint-config/package.json (hunk 14 above)
Comment (rezaerami) [reply packages/tools/eslint-config/package.json:33]: no, it will work both in dependencies and devDependencies
the reason I have put them in devDependencies is just that they are dev packages

File: packages/tools/eslint-config/profile/lib.js (hunk 15 above)
Comment (rezaerami) [reply packages/tools/eslint-config/profile/lib.js:7]: done

File: packages/tools/eslint-config/profile/react.js (hunk 16 above)
Comment (rezaerami) [reply packages/tools/eslint-config/profile/react.js:30]: done

File: packages/tools/kda-cli/.eslintrc.cjs (hunk 17 above)
Comment (rezaerami) [reply packages/tools/kda-cli/.eslintrc.cjs:9]: yes, the reason is in kda-cli we are addressing ts files with .js format which I don't know the reason but when I try to remove the format it doesn't build it anymore!
for example this is an import within kda-cli
`import { IAnswers, IQuestion } from '../questions/questions.js';`

even though the extension for questions file is `.ts`

because of that all the imports will have unresolved warning, so I turned it off for this module

File: packages/libs/react-components/package.json (hunk 13 above)
Comment (rezaerami) [reply packages/libs/react-components/package.json:60]: it's unused, removed

File: packages/tools/eslint-config/profile/lib.js
Comment (rezaerami) [reply packages/tools/eslint-config/profile/lib.js:1]: i will do that in a different PR since it'll effect all package jsons

File: packages/apps/transfer/.eslintrc.js (hunk 10 above)
Comment (rezaerami) [reply packages/apps/transfer/.eslintrc.js:7]: yes this is totally fine since we added tests on purpose to eslint ignore just for transfer since we had some difficulties, but now they are resolved and there's no need to ignore tests anymore, just like other next setups in monorepo

File: packages/apps/transfer/src/hooks/use-did-update-effect.ts [hunk 18]
```diff
@@ -1,6 +1,7 @@
 import { useEffect, useRef } from 'react';
 
-export default (fn: Function, inputs: any[]) => {
+// eslint-disable-next-line @typescript-eslint/ban-types
+export default (fn: Function, inputs: unknown[]): boolean => {
```
Comment (alber70g) [inline packages/apps/transfer/src/hooks/use-did-update-effect.ts:4]: We can type this easily as `() => void`.

```suggestion
export default (fn: () => void, inputs: unknown[]): boolean => {
```

**Review by alber70g — APPROVED** [review]

File: packages/apps/transfer/src/hooks/use-did-update-effect.ts (hunk 18 above)
Comment (rezaerami) [reply packages/apps/transfer/src/hooks/use-did-update-effect.ts]: done

File: packages/tools/eslint-config/package.json (hunk 14 above)
Comment (alber70g) [reply packages/tools/eslint-config/package.json:33]: Yes but for the eslint-config they're actual dependencies, as you install `@kadena/eslint-config` for the configuration specifically
